services/src/tasks/mod.rs

```diff
@@ -0,0 +1,141 @@
+mod error;
+mod in_memory;
+
+use crate::error::Result;
+pub use error::TaskError;
+use geoengine_datatypes::{error::ErrorSource, identifier};
+pub use in_memory::{InMemoryTaskDb, InMemoryTaskDbContext};
+use serde::{Deserialize, Serialize, Serializer};
+use std::{fmt, sync::Arc};
+
+/// A database that allows registering and retrieving tasks.
+#[async_trait::async_trait]
+pub trait TaskDb<C: TaskContext>: Send + Sync {
+    async fn register(&self, task: Box<dyn Task<C>>) -> Result<TaskId, TaskError>;
+
+    async fn status(&self, task_id: TaskId) -> Result<TaskStatus, TaskError>;
+
+    async fn list(&self, options: TaskListOptions) -> Result<Vec<TaskStatusWithId>, TaskError>;
+}
+
+identifier!(TaskId);
+
+/// A task that can run asynchronously and reports its status.
+#[async_trait::async_trait]
+pub trait Task<C: TaskContext>: Send + Sync {
+    async fn run(self: Box<Self>, ctx: C) -> Result<Box<dyn TaskStatusInfo>, Box<dyn ErrorSource>>;
+
+    fn boxed(self) -> Box<dyn Task<C>>
+    where
+        Self: Sized + 'static,
+    {
+        Box::new(self)
+    }
+}
+
+/// A way to supply status updates from a `Task` to a `TaskDb`.
+#[async_trait::async_trait]
+pub trait TaskContext: Send + Sync {
+    /// Set the completion percentage (%) of the task.
+    /// This is a number between 0 and 100.
+    ///
+    /// Moreover, set a status message.
+    ///
+    async fn set_completion(&self, pct_complete: u8, status: Box<dyn TaskStatusInfo>);
```

Comment:
why is this a trait method on the context? shouldn't the task itself set its own status? Who else could know about its completion?

Reply:
It works this way. The task gets the context and can set the completion… or… do you mean something different?